### PR TITLE
[DTensor] Make loss parallel support multi-dimensional mesh

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* #182730
* #183629
* __->__ #180936
* #180937

Extend loss_parallel() from 1D-only to N-D mesh. The TP dim is now any
single mesh dim that shards the class dimension (not necessarily last);
the remaining mesh dims may be Shard (DP/CP) or Replicate.

Output placements by reduction:
  - "none": inherits target placements (Shard on batch dims, Replicate on TP).
  - "sum": Replicate on TP (custom all-reduce makes each rank's value
    globally correct), Partial("sum") on other Shard dims so the cross-rank
    reduction happens lazily on materialization/redistribution.
  - "mean": rejected on N-D mesh — total_weight is a per-rank local count,
    so aggregation across non-TP dims would not yield the correct global
    mean; use "sum"/"none" and divide by the global count.

For reduction="none", the backward handler redistributes grad_output to
the forward output placements, since autograd may deliver it replicated
even when the forward output was batch-sharded.

### CODE REVIEW

**Comment by joecummings** [comment]
Ignore review: thought this was in torchtitan, not pytorch/pytorch

File: torch/distributed/tensor/parallel/loss.py [hunk 1]
```diff
@@ -99,6 +152,17 @@ def _cast_to_dtensor(
         else:
             raise RuntimeError(f"Expected {placements} but got {tensor.placements}.")
     elif isinstance(tensor, torch.Tensor):
+        if any(p.is_shard() for p in placements):
+            # A plain torch.Tensor with Shard placements is ambiguous: it could
+            # be a full global tensor the user expects us to shard, or an
+            # already-local slice. Require a DTensor so the intent is explicit.
```
Comment (tianyu-l) [inline torch/distributed/tensor/parallel/loss.py:158]: This concern doesn't sound absolutely necessary, but I forgot the details on plain tensor inputs. If you don't think this is thoroughly tested and mostly not used, I'm OK with raising.

File: torch/distributed/tensor/parallel/loss.py [hunk 2]
```diff
@@ -456,10 +544,22 @@ def _nll_loss_backward_handler(
     if weight is not None:
         weight = _cast_to_dtensor(weight, all_replicate_placements, spec.mesh)
 
+    # For reduction="none", autograd may deliver grad_output with different
+    # placements than the forward output (e.g. Replicate on a batch-sharded dim),
+    # causing a shape mismatch with the local batch. Redistribute it to match
```
Comment (tianyu-l) [inline torch/distributed/tensor/parallel/loss.py:549]: wait how is this possible ... do you have an example?

**Review by tianyu-l — COMMENTED** [review]
> For reduction="none", the backward handler redistributes grad_output to
the forward output placements, since autograd may deliver it replicated
even when the forward output was batch-sharded.

would like to see an example on this

File: torch/distributed/tensor/parallel/loss.py (hunk 1 above)
Comment (fegin) [reply torch/distributed/tensor/parallel/loss.py:158]: It is real, but not in TorchTitan as if we have the full dtensor, we don't have plan tensors. So this won't affect TorchTitan but in general it is better to raise.

File: torch/distributed/tensor/parallel/loss.py (hunk 2 above)
Comment (fegin) [reply torch/distributed/tensor/parallel/loss.py:549]: With DP in the sharding spec:
```
with loss_parallel():
    loss = F.cross_entropy(dist_logits, dist_target, reduction="none")  # [DP, TP] == [Shard(0), Shard(1)]
    loss.sum().backward()
```

DTensor will propagate the Replicate (it will always choose Replicate along this particular backward path), so grad_output will be `[Replicate, Replicate]` and will mismatch with the target_placements, `[Shard(0), Replicate]` and cause local_tensor shape mismatch.

Comment (tianyu-l) [reply torch/distributed/tensor/parallel/loss.py:549]: It sounds not the root cause. The root cause seems happening before this nll_loss_backward -- `grad_out` shouldn't be created as `Replicate` in the first place, given a `Shard(0)` inputs. Please leave a TODO for further investigation.

**Review by tianyu-l — APPROVED** [review]
